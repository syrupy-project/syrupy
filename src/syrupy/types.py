from typing import (
    Any,
    Callable,
    Hashable,
    Optional,
    Tuple,
    Type,
    Union,
)

SerializableData = Any
SerializedData = Union[str, bytes]
PropertyName = Hashable
PropertyValueType = Type[SerializableData]
PropertyPathEntry = Tuple[PropertyName, PropertyValueType]
PropertyPath = Tuple[PropertyPathEntry, ...]
try:
    from mypy_extensions import NamedArg

    PropertyMatcher = Callable[
        [
            NamedArg(SerializableData, "data"),  # noqa: F821
            NamedArg(PropertyPath, "path"),  # noqa: F821
        ],
        Optional[SerializableData],
    ]
    PropertyFilter = Callable[
        [
            NamedArg(PropertyName, "prop"),  # noqa: F821
            NamedArg(PropertyPath, "path"),  # noqa: F821
        ],
        bool,
    ]

except ImportError:
    globals()["PropertyMatcher"] = Callable[..., Optional[SerializableData]]
    globals()["PropertyFilter"] = Callable[..., bool]
