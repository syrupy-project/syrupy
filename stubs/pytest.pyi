from typing import (
    Any,
    Callable,
    TypeVar,
)

ReturnType = TypeVar("ReturnType")

def fixture(func: Callable[..., ReturnType]) -> Callable[..., ReturnType]: ...

class Function: ...
class Item: ...
