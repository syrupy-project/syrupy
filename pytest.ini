[pytest]
addopts = -p pytester -p syrupy
