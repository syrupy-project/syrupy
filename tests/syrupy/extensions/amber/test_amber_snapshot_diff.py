from types import MappingProxyType

import pytest


def test_snapshot_diff(snapshot):
    my_dict = {
        "field_0": True,
        "field_1": "no_value",
        "nested": {
            "field_0": 1,
        },
    }
    assert my_dict == snapshot
    my_dict["field_1"] = "yes_value"
    assert my_dict == snapshot(diff=0)
    my_dict["nested"]["field_0"] = 2
    assert my_dict == snapshot(diff=0)


def test_snapshot_diff_id(snapshot):
    my_dict = {
        "field_0": True,
        "field_1": "no_value",
        "field_2": 0,
        "field_3": None,
        "field_4": 1,
        "field_5": False,
        "field_6": (True, "hey", 2, None),
        "field_7": {False, "no", 0, None},  # noqa: disable=B033
    }
    dictLargeSnapshot = {
        **my_dict,
        "nested_0": dict(my_dict),
        "nested_1": dict(my_dict),
    }

    assert dictLargeSnapshot == snapshot(name="large snapshot")
    dictDiffLargeSnapshot = {
        **dictLargeSnapshot,
        "field_1": "yes_value",
        "field_6": ("hey", 2, False),
        "field_7": {"yes", 0, None},
    }
    assert dictDiffLargeSnapshot == snapshot(diff="large snapshot")
    dictCase3 = {
        **dictLargeSnapshot,
        "nested_0": MappingProxyType({**my_dict, "field_0": 2}),
        "nested_1": MappingProxyType({**my_dict, "field_0": 2}),
    }
    assert dictCase3 == snapshot(name="case3", diff="large snapshot")


def test_snapshot_no_diff_raises_exception(snapshot):
    my_dict = {
        "field_0": "value_0",
    }
    with pytest.raises(AssertionError, match="SnapshotDoesNotExist"):
        assert my_dict == snapshot(diff="does not exist index")
