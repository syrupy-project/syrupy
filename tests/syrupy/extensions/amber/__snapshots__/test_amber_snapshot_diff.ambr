# serializer version: 1
# name: test_snapshot_diff
  dict({
    'field_0': True,
    'field_1': 'no_value',
    'nested': dict({
      'field_0': 1,
    }),
  })
# ---
# name: test_snapshot_diff.1
      ...
  -   'field_1': 'no_value',
  +   'field_1': 'yes_value',
      ...
# ---
# name: test_snapshot_diff.2
      ...
  -   'field_1': 'no_value',
  +   'field_1': 'yes_value',
      ...
  -     'field_0': 1,
  +     'field_0': 2,
    ...
# ---
# name: test_snapshot_diff_id.1
      ...
  -   'field_1': 'no_value',
  +   'field_1': 'yes_value',
      ...
  -     True,
        ...
  -     None,
  +     False,
      ...
  -     'no',
  +     'yes',
  -     False,
  +     0,
      ...
# ---
# name: test_snapshot_diff_id[case3]
        ...
  -   'nested_0': dict({
  +   'nested_0': mappingproxy({
  -     'field_0': True,
  +     'field_0': 2,
          ...
  -   'nested_1': dict({
  +   'nested_1': mappingproxy({
  -     'field_0': True,
  +     'field_0': 2,
          ...
# ---
# name: test_snapshot_diff_id[large snapshot]
  dict({
    'field_0': True,
    'field_1': 'no_value',
    'field_2': 0,
    'field_3': None,
    'field_4': 1,
    'field_5': False,
    'field_6': tuple(
      True,
      'hey',
      2,
      None,
    ),
    'field_7': set({
      'no',
      False,
      None,
    }),
    'nested_0': dict({
      'field_0': True,
      'field_1': 'no_value',
      'field_2': 0,
      'field_3': None,
      'field_4': 1,
      'field_5': False,
      'field_6': tuple(
        True,
        'hey',
        2,
        None,
      ),
      'field_7': set({
        'no',
        False,
        None,
      }),
    }),
    'nested_1': dict({
      'field_0': True,
      'field_1': 'no_value',
      'field_2': 0,
      'field_3': None,
      'field_4': 1,
      'field_5': False,
      'field_6': tuple(
        True,
        'hey',
        2,
        None,
      ),
      'field_7': set({
        'no',
        False,
        None,
      }),
    }),
  })
# ---
